---
title: "CFHCyear1"
author: "Katherine Antosca"
date: "11/27/2018"
output: word_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#load necessary libraries
library(dada2)
library(ShortRead)
library(ggplot2)
library(breakaway)
library(DivNet)
library(corncob)
library(magrittr)
library(dplyr)
library(vegan)
library(phyloseq)
library(ade4)
library(car)
library(phangorn)
library(multcomp)
library(RColorBrewer)
library(GUniFrac)
library(pheatmap)
library(prettyPCoA)
library(ggsignif)
library(DECIPHER)
```




```{r alpha, echo = FALSE}
#read in sample data
sampdf <- read.csv("ClinData.csv")
rownames(sampdf) <- sampdf$mblid

#Remove CF samples after the 1 year time point
year1seqtab <- seqtab.nochim[which(rownames(seqtab.nochim) %in% rownames(sampdf)),]


#Check row order
sampdf <- sampdf[match(rownames(year1seqtab), sampdf$mblid),]
rownames(sampdf) == rownames(year1seqtab)

#convert to phyloseq object

phylo <- phyloseq(otu_table(year1seqtab, taxa_are_rows=FALSE), 
               sample_data(sampdf), 
               tax_table(taxa)) 

plot_richness(phylo, x = "TimePeriod", measures = "Shannon", color = "Status")

alphadiv <- estimate_richness(phylo)
#extract shannon diversity and add it to sampdf
Shannon <- alphadiv$Shannon
Simpson <- alphadiv$Simpson
sampdf <- cbind(sampdf, Shannon, Simpson)
#Set order for plotting by age and Status
sampdf$TimePeriod <- factor(sampdf$TimePeriod, levels = c("6W","4M","6M","9M","12M"))
sampdf$Status <- factor(sampdf$Status, levels = c("HC","CF"))
Greens <- brewer.pal(5,"Greens")
Blues <- brewer.pal(5,"Blues")
GreenBlues <- as.vector(rbind(Greens, Blues))

```  

```{r HCalphaplot, echo = FALSE}
HCalphaplot <- ggplot(data = sampdf[sampdf$Status=="HC",], aes(x = TimePeriod, y = Shannon, 
                                                               group = TimePeriod)) + 
  geom_boxplot(fill = Greens, color = "darkgreen") + 
  geom_signif(comparisons = list(c("6W", "12M")), map_signif_level=TRUE) +
  theme_bw(base_size = 25) +
  xlab("Age in Months") +
  scale_x_discrete(labels = c("1M","4M","6M","9M","12M"))
HCalphaplot
pdf(file = "ShannonHC.pdf",height = 8, width = 12)
  HCalphaplot
dev.off()
```
```{r CFalphaplot, echo = FALSE}
CFalphaplot <- ggplot(data = sampdf[sampdf$Status=="CF",], aes(x = TimePeriod, y = Shannon, group = TimePeriod)) + geom_boxplot(fill = Blues, color = "mediumblue") +
  geom_signif(comparisons = list(c("6W", "12M")), map_signif_level=TRUE) +
  theme_bw(base_size = 25) +
  xlab("Age in Months") +
  scale_x_discrete(labels = c("1M","4M","6M","9M","12M"))
CFalphaplot
pdf(file = "ShannonCF.pdf",height = 8, width = 12)
  CFalphaplot
dev.off()
```

```{r alphaplot, echo = FALSE}
alphaplot <- ggplot(data = sampdf, aes(x = TimePeriod, y = Shannon, color = Status)) + 
  geom_boxplot(fill = GreenBlues) +
  scale_color_manual(values = c("darkgreen", "mediumblue"), 
                     labels = c("General Population", "Cystic Fibrosis"))+
  theme_bw(base_size = 25) +
  xlab("Age in Months") +
  scale_x_discrete(labels = c("1M","4M","6M","9M","12M"))
alphaplot
pdf(file = "ShannonAll.pdf",height = 8, width = 14)
  alphaplot
dev.off()
  

```


```{r phylotree, echo = FALSE}
#multiple alignments with DECIPHER then create the tree with phangorn
seqs <- getSequences(year1seqtab)
# This propagates to the tip labels of the tree
names(seqs) <- seqs 
alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA,verbose=FALSE)

#From https://bioconductor.org/help/course-materials/2017/BioC2017/Day1/Workshops/Microbiome/MicrobiomeWorkflowII.html The phangorn R package is then used to construct a phylogenetic tree. Here we first construct a neighbor-joining tree, and then fit a GTR+G+I (Generalized time-reversible with Gamma rate variation) maximum likelihood tree using the neighbor-joining tree as a starting point.

phangAlign <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phangAlign)
treeNJ <- NJ(dm) # Note, tip order != sequence order
fit = pml(treeNJ, data=phangAlign)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
        rearrangement = "stochastic", control = pml.control(trace = 0))

```


```{r beta, echo = FALSE}


```












